# Overview
We want to maximise $n/\phi(n)$ subject to $n \le 1,000,000$.

The Euler phi function is multiplicative, which means that $\phi(mn) = \phi(m)\phi(n)$ whenever $m$ and $n$ are coprime. We can use this fact to make the search easier.

Suppose $n = p^k$ is a prime power. Then $\phi(n) = p^{k-1}(p-1)$, so $n/\phi(n) = p/(p-1)$. Hence, it is only worth considering $k = 1$. Since $\phi$ is multiplicative, this means that for general $n$, we only need to consider those numbers that are a product of distinct primes. But $p/(p-1)$ is a decreasing function of $p$, so choosing small primes will maximise $n/\phi(n)$ while minimising $n$.

Therefore, we only need to consider products of the smallest, distinct primes, i.e. the primorial numbers $p_k\#$.

In [2]:
from functools import reduce

# Hardcode the first few primes
primes = [2, 3, 5, 7, 11, 13, 17, 19]
for i in range(len(primes)):
    prefix = primes[:i+1]
    primorial = reduce(lambda x, y: x * y, prefix)
    print(f"Product of {prefix}: {primorial}")

Product of [2]: 2
Product of [2, 3]: 6
Product of [2, 3, 5]: 30
Product of [2, 3, 5, 7]: 210
Product of [2, 3, 5, 7, 11]: 2310
Product of [2, 3, 5, 7, 11, 13]: 30030
Product of [2, 3, 5, 7, 11, 13, 17]: 510510
Product of [2, 3, 5, 7, 11, 13, 17, 19]: 9699690


We can see that $$17\# = 510,510 < 1,000,000$$ and $$19\# = 9,699,690 > 1,000,000,$$ so the answer is $510,510$.